content/posts/2025-09-01-gsoc-ubsan-trap-messages.md

```diff
@@ -0,0 +1,410 @@
+trap:                                             ; preds = %entry
+  call void @llvm.ubsantrap(i8 0) #4, !dbg !31, !nosanitize !21
+  unreachable, !dbg !31, !nosanitize !21
+cont:                                             ; preds = %entry
+  ret i32 %3, !dbg !34
+}
+;...
+!29 = !DILocation(line: 5, column: 13, scope: !17)
+!30 = !{!"branch_weights", i32 1048575, i32 1}
+!31 = !DILocation(line: 0, scope: !32, inlinedAt: !29)
+!32 = distinct !DISubprogram(name: "__clang_trap_msg$Undefined Behavior Sanitizer$Integer addition overflowed", scope: !2, file: !2, type: !33, flags: DIFlagArtificial, spFlags: DISPFlagDefinition, unit: !14)
+```
+
+
+The debug metadata for the `@llvm.ubsantrap` call is `!31`.  That `DILocation` has the scope of the `DISubprogram` assigned to `!32` which is the artificial function which encodes the trap category (`Undefined Behavior Sanitizer`) and the trap message (`Integer addition overflowed`). Note that the `DILocation` for `!31` has `inlinedAt:` which tells us that the trap was inlined from !32 into the location at !29 which is the location of the `a + b` expression in the `add` function.
+
+
+I implemented this change on this [PR](https://github.com/llvm/llvm-project/pull/145967).
+
+
+### Debug info size changes
+
+
+One concern that a reviewer had was the debug info size difference. This was one of the motivations for putting this feature under the new `-fsanitize-debug-trap-reasons` flag because initially (prior to code review), my mentors and I planned to have the trap feature flag accompany the `-fsanitize-trap=` flag. Although the `-fsanitize-debug-trap-reasons` flag is on by default (so long as trapping UBSan is enabled), having the trap reason feature under a flag allows users to opt-out by using the `-fno-sanitize-debug-trap-reasons` flag. 
+
+Using bloaty, I tested a release build of clang with the `-fsanitize-debug-trap-reasons` flag enabled, and one with it disabled (`-fno-sanitize-debug-trap-reasons`). We found that the size difference was negligible; results are below.
+
+
+```
+   FILE SIZE        VM SIZE   
+--------------  --------------
+ +0.3% +6.01Mi  +0.3% +6.01Mi    ,__debug_info
+ +2.0% +2.26Mi  [ = ]       0    [Unmapped]
+ +1.2% +1.35Mi  +1.2% +1.35Mi    ,__apple_names
+ +0.0% +1.01Mi  +0.0% +1.01Mi    ,__debug_str
+ +0.8%  +636Ki  +0.8%  +635Ki    ,__debug_line
+ +0.4%  +161Ki  +0.4%  +161Ki    ,__debug_ranges
+ +0.4% +47.9Ki  +0.4% +47.9Ki    ,__debug_abbrev
+ +0.0%     +14  +0.0%     +14    ,__apple_types
+ [ = ]       0  +0.0%      +8    ,__common
+ [ = ]       0  +7.1%      +4    ,__thread_bss
+ -0.0%      -4  -0.0%      -4    ,__const
+ -0.0% -1.27Ki  -0.0% -1.27Ki    ,__cstring
+ +0.2% +11.5Mi  +0.1% +9.19Mi    TOTAL
+ ```
+
+
+Note it is likely the code size difference is negligible because because in optimized builds trap instructions in a function get merged together which causes the additional debug info my patch adds to be dropped.
```

Comment:
I would point out where the debug info size impact would mainly come from: https://github.com/llvm/llvm-project/pull/145967#issuecomment-3068862442

And also mention, the size increase is tied to how many traps get emitted (since we emit a new DW_TAG_subprogram DIE for each trap). I would also link to the comment here: https://github.com/llvm/llvm-project/pull/154618#issuecomment-3225724300, which shows that in production Google saw a pretty large size increase in debug-info with this (~15% total binary size increase IIRC). So this could be a potential future thing to look into. Why does the binary size increase by that much and what could we do about it? As it stands, they had to disable the trap reason because of that, which isn't great.